logstash-core/src/main/java/org/logstash/config/ir/compiler/ComputeStepSyntaxElement.java

```diff
@@ -33,84 +32,98 @@
     private static final ISimpleCompiler COMPILER = new SimpleCompiler();
 
     /**
-     * Sequential counter to generate the class name
+     * Global cache of runtime compiled classes to prevent duplicate classes being compiled.
+     * across pipelines and workers.
      */
-    private static final AtomicLong classSeqCount = new AtomicLong();
+    private static final Map<ComputeStepSyntaxElement<?>, Class<? extends Dataset>> CLASS_CACHE
+        = new HashMap<>(5000);
 
     /**
      * Pattern to remove redundant {@code ;} from formatted code since {@link Formatter} does not
      * remove those.
      */
     private static final Pattern REDUNDANT_SEMICOLON = Pattern.compile("\n[ ]*;\n");
 
+    private static final String CLASS_NAME_PLACEHOLDER = "CLASS_NAME_PLACEHOLDER";
+
+    private static final Pattern CLASS_NAME_PLACEHOLDER_REGEX = Pattern.compile(CLASS_NAME_PLACEHOLDER);
+
     private final Iterable<MethodSyntaxElement> methods;
 
     private final ClassFields fields;
 
     private final Class<T> type;
 
-    private final long classSeq;
+    private final String normalizedSource;
 
     public static <T extends Dataset> ComputeStepSyntaxElement<T> create(
-        final Iterable<MethodSyntaxElement> methods, final ClassFields fields,
-        final Class<T> interfce) {
+        final Iterable<MethodSyntaxElement> methods,
+        final ClassFields fields,
+        final Class<T> interfce)
+    {
         return new ComputeStepSyntaxElement<>(methods, fields, interfce);
     }
 
-    private ComputeStepSyntaxElement(final Iterable<MethodSyntaxElement> methods,
-        final ClassFields fields, final Class<T> interfce) {
+    private ComputeStepSyntaxElement(
+        final Iterable<MethodSyntaxElement> methods,
+        final ClassFields fields,
+        final Class<T> interfce)
+    {
         this.methods = methods;
         this.fields = fields;
         type = interfce;
-        classSeq = classSeqCount.incrementAndGet();
+
+        // normalizes away the name of the class so that two classes of different name but otherwise
+        // equivalent syntax get correctly compared by {@link #equals(Object)}.
+        normalizedSource = generateCode(CLASS_NAME_PLACEHOLDER);
     }
 
     @SuppressWarnings("unchecked")
-    public T instantiate(Class<? extends Dataset> clazz) {
-        try {
-            return (T) clazz.<T>getConstructor(Map.class).newInstance(ctorArguments());
+    public T instantiate() {
+         try {
+            synchronized (COMPILER) {
+                Class<? extends Dataset> clazz = CLASS_CACHE.get(this);
+                if (clazz == null) {
+                    clazz = compile();
+                    CLASS_CACHE.put(this, clazz);
+                }
+                return (T) clazz.getConstructor(Map.class).newInstance(ctorArguments());
+            }
```

Comment:
I tried to use a `ConcurentHashMap` and use `computeIfAbsent` for the `compile()` result but both the compilation **and** `newInstance()` call need to be both in the same exclusive context otherwise there are some race condition issues.

Reply:
do you feel like this might eventually be a bottleneck (given that you sized the class cache map to 5000) ?

Reply:
- Note that this is basically how this was (the global lock) before (in 7.5.1). It is certainly a bottleneck, but ATM I don't see a way around it. For a cross-pipeline and cross-workers cache to be efficient (to globally avoid double-compilations) no 2 compilations should ne performed concurrently. Also, the `cook` call, the `setParentClassLoader` call and the generated class instantiation must be done atomically  (for reasons I do not yet grasp).
- This bottleneck mostly hurts in in the mutli-pipelines scenario, at initial startup where there are a lot more chances of needing to perform an actual compilation. 
- I put in an initial capacity of 5000 because the default initial capacity of 16 is definitely too small, any decent config endup generating tons of classes. 5000 is arbitrary. 
- Not that there is currently a cache eviction leak problem with this code because it will not evict dead classes across pipelines reloads. This problem was already present in 7.5.1 and I decided not to tackle it just now because practically speaking I do not think the cache should grow at a rate where that leak could have an impact. I will document that problem in the code and open a followup issue about it too.

Reply:
resolving this conversation, will followup with an eviction issue.